File: concat_moments.jl
#!/usr/bin/env julia

# Concatenate chunk CSV files without external package dependencies.
# Keeps the header from the first file and appends data rows from all chunks.

using Printf

function main()
    input_dir = "SimulatedData"
    output_path = joinpath(input_dir, "moments.csv")

    all_files = sort(readdir(input_dir; join=true))
    chunk_files = filter(f -> occursin(r"moments_\d+\.csv$", basename(f)), all_files)

    isempty(chunk_files) && error("No chunk files found in SimulatedData/")

    rows_written = 0
    header_line = nothing

    open(output_path, "w") do fout
        for f in chunk_files
            first_line_in_file = true
            open(f, "r") do fin
                for line in eachline(fin)
                    if first_line_in_file
                        if header_line === nothing
                            header_line = line
                            println(fout, line)
                        elseif line != header_line
                            error("Header mismatch in $(f)")
                        end
                        first_line_in_file = false
                        continue
                    end
                    println(fout, line)
                    rows_written += 1
                end
            end
        end
    end

    @printf("Concatenated %d files -> %s (%d rows)\n", length(chunk_files), output_path, rows_written)
end

main()

File: PythonEstimation/concat_moments.py
"""concat_moments.py

Concatenate chunk CSV files produced by the simulate-moments job array.
Keeps the header from the first file and appends data rows from all chunks.
Python translation of concat_moments.jl.
"""

import glob
import os
import sys

INPUT_DIR = "SimulatedData"
OUTPUT_PATH = os.path.join(INPUT_DIR, "moments.csv")


def main():
    chunk_files = sorted(
        glob.glob(os.path.join(INPUT_DIR, "moments_*.csv"))
    )

    if not chunk_files:
        sys.exit(f"No chunk files found in {INPUT_DIR}/")

    header_line = None
    rows_written = 0

    with open(OUTPUT_PATH, "w") as fout:
        for fpath in chunk_files:
            with open(fpath, "r") as fin:
                first = True
                for line in fin:
                    if first:
                        first = False
                        if header_line is None:
                            header_line = line
                            fout.write(line)
                        elif line != header_line:
                            sys.exit(f"Header mismatch in {fpath}")
                        continue
                    fout.write(line)
                    rows_written += 1

    print(
        f"Concatenated {len(chunk_files)} files -> {OUTPUT_PATH} "
        f"({rows_written} rows)"
    )


if __name__ == "__main__":
    main()
